### PR TITLE
Added bit_field intrisic

### PR DESCRIPTION
**Pull request by Eichenherz**
## Description
Added BitField(Un)SignedExtract intrinsic.

### CODE REVIEW

File: include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl [hunk 1]
```diff
@@ -129,6 +129,15 @@ void memoryBarrier(uint32_t memoryScope, uint32_t memorySemantics);
...
 [[vk::ext_instruction(spv::OpBitcast)]]
 T bitcast(U);
+
+template<typename Unsigned>
+[[vk::ext_instruction( spv::OpBitFieldUExtract )]]
+Unsigned bitfieldExtractUnsigned( Unsigned val, uint32_t offsetBits, uint32_t numBits );
+
+template<typename Signed>
+[[vk::ext_instruction( spv::OpBitFieldSExtract )]]
+Signed bitfieldExtractSigned( Signed val, uint32_t offsetBits, uint32_t numBits );
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl:139]: keep the naming 1:1 with SPIR-V but without `Op` prefix, so `bitFieldUExtract` and `bitFieldSExtract`

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 2]
```diff
@@ -102,6 +108,13 @@ uint16_t countl_zero(T n)
     return impl::clz<sizeof(T)*8>(n);
 }
 
+// TODO: templated "switch" depending on signedness  
+template<typename Unsigned>
+Unsigned bitfield_extract_unsigned( Unsigned val, uint32_t offsetBits, uint32_t numBits )
+{
+    return spirv::bitfieldExtractUnsigned<Unsigned>( val, offsetBits, numBits );
+}
+
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:117]: move this to the `glsl` namespace header and conform to
https://registry.khronos.org/OpenGL-Refpages/gl4/html/bitfieldExtract.xhtml

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/bit.hlsl:117]: btw if you want to "template switch" use a struct  and partial specialization as the implementation behind the scenes

Comment (Eichenherz) [reply include/nbl/builtin/hlsl/bit.hlsl]: how do I use bitfieldExtract in hlsl/cpp "common" code ? glsl:: does not implement any cpp functions. Is it ok to do this or do I just move the hlsl that uses it out of the "common" scope ?

File: include/nbl/builtin/hlsl/glsl_compat/core.hlsl [hunk 3]
```diff
@@ -56,6 +56,18 @@ T atomicCompSwap(NBL_REF_ARG(T) ptr, T comparator, T value)
...
 
+template<typename T>
+T bitfieldExtract( T val, uint32_t offsetBits, uint32_t numBits )
+{
+    static_assert( is_integral<T>::value );
+    if( is_unsigned<T>::value )
+    {
+        return spirv::bitFieldUExtract<T>( val, offsetBits, numBits );
+    }
+    //if( is_signed<T>::value )
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/glsl_compat/core.hlsl:67]: else ?

File: include/nbl/builtin/hlsl/glsl_compat/bit.hlsl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            static_assert( false, "T must be an integral type" );
+        }
+        if( isSigned )
+        {
+            return spirv::bitFieldSExtract<T>( val, offsetBits, numBits );
+        }
+        return spirv::bitFieldUExtract<T>( val, offsetBits, numBits );
+    }
+};
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/glsl_compat/bit.hlsl:36]: thats not how partial specialization works

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/glsl_compat/bit.hlsl:36]: also the impl goes into `glsl_compat/core.hlsl`

File: include/nbl/builtin/hlsl/glsl_compat/core.hlsl [hunk 5]
```diff
@@ -83,6 +83,50 @@ void tess_ctrl_barrier() {
...
+        return spirv::bitFieldUExtract<T>( val, offsetBits, numBits );
+    } 
+};
+
+}
+
+template<typename T>
+T bitfieldExtract( T val, uint32_t offsetBits, uint32_t numBits )
+{
+    return impl::bitfieldExtract<T, is_signed<T>::value, is_integral<T>::value>( val, offsetBits, numBits );
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/glsl_compat/core.hlsl:127]: did you actually test your code though, cause I can see with my very own eyes this is invalid syntax.

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 2 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/bit.hlsl]: create an alias a for the GLM function that does in the `ifndef __HLSL_VERSION` `else` case

File: include/nbl/builtin/hlsl/bit.hlsl [hunk 6]
```diff
@@ -3,6 +3,7 @@
 
 
 #include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/type_traits.hlsl>
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/bit.hlsl:6]: you're not using it here so don't add it

File: include/nbl/builtin/hlsl/bit.hlsl (hunk 2 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/bit.hlsl]: ok just provide the C++ alias, and we can merge

**Pull request merged**